design/thin-edge-actors-design.md

```diff
@@ -553,23 +583,273 @@ impl From<Msg2> for Msg {
 
 #### Channel types
 
-#### Channel creation and ownership
+We don't want to expose that the messages are sent over `mpsc` channels. 
+Abstracting both the senders and receivers, we can implement specific mechanisms
+to manage message priority, timeout, cancellation and more.
 
-#### Actor with no inputs
+The receiver part is abstracted by a concrete type: a mailbox for a specific type of messages.
+A key design point is that *all* the interactions with an actor must go through such a mailbox,
+including runtime errors, cancellations and timeouts.
+Doing so simplifies the actor event loops to `while let Some(message) = self.mailbox.next().await { ... }`
+and gives the flexibility to improve the system with new message delivery mechanisms.
 
-#### Out-of-band runtime messages
+```rust
+/// A mailbox that gather *all* the messages sent to an actor, including runtime messages.
+impl<M : From<RuntimeMessage>> Mailbox<M> {
+  /// Pop from the mailbox the message with the highest priority
+  /// 
+  /// Await till a messages is available.
+  /// Return `None` when all the senders to this mailbox have been dropped and all the messages consumed.
+  pub async fn next(&mut self) -> Option<M> {
+    // ...
+  }
+  
+  /// Turn on/off logging of the messages consumed from this mailbox.
+  /// 
+  /// The messages are logged when returned by the `next()` method.
+  pub fn log_inputs(&mut self, on: bool) {
+    // ...
+  }
+}
 
-#### Size of the channel buffers
+pub enum RuntimeMessage {
+  RuntimeError(RuntimeError),
+  Timeout,
+  Shutdown,
+  LogInputs(bool),
+  LogOutputs(bool),
+  // ...
+}
+```
 
-### Behavior
+The message sender of an actor is implemented as an `Address<M>`, the address of its mailbox.
 
-### Instantiation
+```rust
+/// Create a new mailbox with its address
+pub fn new_mailbox<M>(buffer: usize) -> (Mailbox<M>, Address<M>) {
+  // ...
+}
 
-### Discovery
+impl<M : From<RuntimeMessage>> Address<M> {
+  /// Send a message to the mailbox with that address
+  /// 
+  /// Await the messages is actually in the mailbox.
+  /// Fail when the mailbox has been dropped.
+  pub async fn send_message(&mut self, message: M) -> Result<(), ChannelError> {
+    // ...
+  }
 
-### Runtime
+  /// Clone this address which can then be used from another actor
+  fn clone(&self) -> Self {
+    // ...
+  }
+}
+```
+
+However, actors can not directly use the addresses of their peers.
+If an actor sends messages of type `A` to an actor that processes messages of type `B: From<A>`,
+then we do not want the source actor to be aware of the `B` message type,
+and still less the unrelated message types consumed by the target as `C: Into<B>`.
+Indeed, that `B` message type encompasses all the message kinds supported by the target,
+and these, as the `C` type, can be completely unrelated to the source business domain that is described by the `A` type.
```

Comment:
It was hard to follow those couple of sentences.

Reply:
When I feel the urgent need to add a diagram this is a sign that this is not so clear even to me ;-).

I will try to come with something clearer.